File: packages/bridge/src/remote.ts
import { invoke } from './invoke.js';

/** One bundle an update offers. */
export interface BundleUpdate {
  /** The name of the bundle. */
  name: string;
  /** The version of the bundle. */
  version: string;
  /**
   * Where the bundle is downloaded from. When absent the client falls back to the remote's
   * default url (`GET /bundles/:name/:version`).
   */
  downloadUrl?: string;
  /** Integrity value to verify the downloaded bundle. */
  integrity?: string;
  /** Arbitrary string-valued metadata for this bundle. */
  metadata?: Record<string, string>;
}

/** The update document the remote server serves. */
export interface Update {
  /** The unique id of this update. */
  id: string;
  /** When the update was created, formatted according to ISO 8601. */
  createdAt: string;
  /** The update model version this update is written against. */
  runtimeVersion: number;
  /** The bundles this update offers. */
  bundles: BundleUpdate[];
  /** Arbitrary string-valued metadata for this update. */
  metadata: Record<string, string>;
}

/** The signature the server sent for an update, parsed from the `wvb-signature` header. */
export interface UpdateSignature {
  /** The id of the key the update was signed with. */
  keyId: string;
  /** The signature value. */
  sig: string;
  /** The signature algorithm. */
  alg: string;
}

export interface RemoteUpdateResponse {
  /** Update information parsed from the response body. */
  update: Update;
  /** "ETag" header value, to pass back on the next request. */
  etag?: string;
  /** Signature information for this update. */
  signature?: UpdateSignature;
}

export interface RemoteGetUpdateOptions {
  /** The etag of the update previously received; sent as "if-none-match". */
  etag?: string;
  /** Release channel to fetch the update from. */
  channel?: string;
}

/**
 * Fetches the update document. Resolves `null` when the server answered `304 Not Modified`,
 * i.e. the update matching `etag` is still the current one.
 */
async function getUpdate(options?: RemoteGetUpdateOptions): Promise<RemoteUpdateResponse | null> {
  return invoke<RemoteUpdateResponse | null>('remoteGetUpdate', { options });
}

/** Downloads `url` into `filepath`. The file is not staged, so nothing is served from it yet. */
async function download(url: string, filepath: string): Promise<void> {
  return invoke<void>('remoteDownload', { url, filepath });
}

export interface RemoteApi {
  getUpdate: typeof getUpdate;
  download: typeof download;
}

export const remote: RemoteApi = {
  getUpdate,
  download,
};

File: packages/tauri/src/command/remote.rs
use crate::WebviewBundleExtra;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::PathBuf;
use tauri::{AppHandle, Runtime, command};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BundleUpdate {
  pub name: String,
  pub version: String,
  pub download_url: Option<String>,
  pub integrity: Option<String>,
  pub metadata: Option<HashMap<String, String>>,
}

impl From<wvb::remote::BundleUpdate> for BundleUpdate {
  fn from(value: wvb::remote::BundleUpdate) -> Self {
    Self {
      name: value.name,
      version: value.version,
      download_url: value.download_url,
      integrity: value.integrity,
      metadata: value.metadata,
    }
  }
}

impl From<BundleUpdate> for wvb::remote::BundleUpdate {
  fn from(value: BundleUpdate) -> Self {
    Self {
      name: value.name,
      version: value.version,
      download_url: value.download_url,
      integrity: value.integrity,
      metadata: value.metadata,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Update {
  pub id: String,
  pub created_at: String,
  pub runtime_version: u8,
  pub bundles: Vec<BundleUpdate>,
  pub metadata: HashMap<String, String>,
}

impl From<wvb::remote::Update> for Update {
  fn from(value: wvb::remote::Update) -> Self {
    Self {
      id: value.id,
      created_at: value.created_at,
      runtime_version: value.runtime_version,
      bundles: value.bundles.into_iter().map(BundleUpdate::from).collect(),
      metadata: value.metadata,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSignature {
  pub key_id: String,
  pub sig: String,
  pub alg: String,
}

impl From<wvb::remote::UpdateSignature> for UpdateSignature {
  fn from(value: wvb::remote::UpdateSignature) -> Self {
    Self {
      key_id: value.key_id,
      sig: value.sig,
      alg: value.alg,
    }
  }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteUpdateResponse {
  pub update: Update,
  pub etag: Option<String>,
  pub signature: Option<UpdateSignature>,
}

impl From<wvb::remote::RemoteUpdateResponse> for RemoteUpdateResponse {
  fn from(value: wvb::remote::RemoteUpdateResponse) -> Self {
    Self {
      update: value.update.into(),
      etag: value.etag,
      signature: value.signature.map(UpdateSignature::from),
    }
  }
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RemoteGetUpdateOptions {
  pub etag: Option<String>,
  pub channel: Option<String>,
}

impl From<RemoteGetUpdateOptions> for wvb::remote::RemoteGetUpdateOptions {
  fn from(value: RemoteGetUpdateOptions) -> Self {
    let mut options = wvb::remote::RemoteGetUpdateOptions::default();
    if let Some(etag) = value.etag {
      options = options.etag(etag);
    }
    if let Some(channel) = value.channel {
      options = options.channel(channel);
    }
    options
  }
}

#[command]
pub async fn remote_get_update<R: Runtime>(
  app: AppHandle<R>,
  options: Option<RemoteGetUpdateOptions>,
) -> crate::Result<Option<RemoteUpdateResponse>> {
  let wvb = app.wvb();
  let resp = wvb
    .remote()
    .ok_or(crate::Error::RemoteIsNotInitialized)?
    .get_update(options.map(Into::into))
    .await?
    .map(RemoteUpdateResponse::from);
  Ok(resp)
}

#[command]
pub async fn remote_download<R: Runtime>(
  app: AppHandle<R>,
  url: String,
  filepath: PathBuf,
) -> crate::Result<()> {
  let wvb = app.wvb();
  wvb
    .remote()
    .ok_or(crate::Error::RemoteIsNotInitialized)?
    .download(url, &filepath, None)
    .await?;
  Ok(())
}
